### PR TITLE
fix(core): rank CBP mature units by bias-corrected utility in replacement selection

### LINKED ISSUE #2343: CBP replaces units by the raw utility EMA, not the paper's bias-corrected utility (Dohare et al. 2024, Eq. 8) — it re-replaces its own freshly reset units
## Summary

`alberta_framework/core/continual_backprop.py::_select_replacement_index` ranks mature units by the **raw** utility EMA. The algorithm it cites ranks by the **bias-corrected** EMA.

Dohare, Hernandez-Garcia, Lan, Rahman, Mahmood & Sutton (2024), *"Loss of plasticity in deep continual learning"*, **Nature 632, 768–774**, Eq. 7–8:

```
u[l,i,t]     = eta * u[l,i,t-1] + (1 - eta) * y[l,i,t]        (7)
u_hat[l,i,t] = u[l,i,t] / (1 - eta ** a[l,i,t])               (8)
```

`a[l,i,t]` is the **per-unit age**. Algorithm 1 then selects the eligible units "with the smallest utility", i.e. smallest `u_hat`.

The authors' released implementation agrees — https://github.com/shibhansh/loss-of-plasticity, `lop/algos/gnt.py`:

```python
# GnT.update_utility
bias_correction = 1 - self.decay_rate ** self.ages[layer_idx]
...
self.bias_corrected_util[layer_idx] = self.util[layer_idx] / bias_correction

# GnT.test_features
new_features_to_replace = torch.topk(-self.bias_corrected_util[i][eligible_feature_indices],
                                     num_new_features_to_replace)[1]
```

`lop/algos/cbp_linear.py` and `lop/algos/cbp_conv.py` use the same per-unit-age correction.

## Why it does not cancel

The correction is per unit, and `_replace_one_unit` sets a replaced unit's `age` **and** `utility` back to 0. So eligible units in the same layer genuinely differ in age and therefore in warm-up bias. At the module defaults (`decay_rate=0.99`, `maturity_threshold=100`) a unit that has just become eligible reads at `1 - 0.99**100 = 0.6340` of its true utility, while a long-lived unit reads at ~1.0. Ranking on the raw EMA therefore systematically targets the youngest eligible units — precisely the units the maturity threshold exists to protect. The paper states the motivation explicitly:

> "initializing the outgoing weight to zero makes the new unit vulnerable to immediate reinitialized as it has zero utility. To protect new units from a reinitialization for maturity threshold m number of updates."

The maturity threshold only defers the problem by `m` steps; without Eq. 8 the freshly reset unit is still the lowest-ranked unit the moment it matures.

## Falsifier (deterministic, pristine `main` @ `c9aba7b5`)

Both scenarios use `decay_rate=0.99`, `maturity_threshold=100`, and drive the real `update_utility` EMA.

**A — young-strong vs old-weak.** Unit 0 has seen a constant contribution of `1.0` for 100 steps; unit 1 a constant `0.9` for 700 steps.

```
raw EMA        : unit0=0.633967 (age 100)   unit1=0.899209 (age 700)
Eq. 8 corrected: unit0=1.000000             unit1=0.900001
true steady-state contributions: unit0=1.0  unit1=0.9

main            REPLACED UNIT: [0]   <- the strictly more useful unit
Eq. 8 / gnt.py  REPLACED UNIT: [1]
```

**B — repeat victim.** Four units with *identical* true contribution `1.0`; unit 2 was reset 100 steps ago (i.e. it is the unit CBP replaced last time and has only just matured).

```
raw EMA        : [0.999997, 0.999997, 0.633967, 0.999997]
Eq. 8 corrected: [0.999997, 0.999997, 1.000000, 0.999997]

main            REPLACED UNIT: [2]   <- re-replaces its own fresh unit
Eq. 8 / gnt.py  REPLACED UNIT: [0]
```

## Scope

- `core/continual_backprop.py` is **not** a registered evidence source (`alberta_framework/evaluation/*` registers `average_reward`, `baseline_optimizers`, `canonical_upgd`, `ftl_world_model`, `intelligence_amplification`, `interaction_features`, `normalizers`, `oak`, `options`, `optimizers`, `types`, `update_safety` — not this file). `alberta-evidence-status` reports byte-identical per-claim validity before and after.
- `benchmarks/ipmnist_screening.py` carries its own CBP implementation and does not import `core.continual_backprop`, so no screening shard, summary, or incumbent number moves.
- In-tree consumers are the public `alberta_framework` API (the robot track's `core/{...,continual_backprop,...}` import) and the tests.

## Precedent

Same defect class as #480 (`fix(features): bias-correct EMA utility outputs`, merged) and issue #431, which fixed the warm-up debias for `future_utility` / `feature_discovery`. `core/continual_backprop.py` was not covered by that change.

## Not changed here (separate, smaller)

Eligibility uses `age >= maturity_threshold`; the paper says "units with age **more than** m" and `gnt.py` uses `torch.where(self.ages > self.maturity_threshold)`. That is a one-step off-by-one and a separate variable; deliberately left out of the fix so this change has exactly one moving part.

### PR DESCRIPTION
**Pull request by Kayvan-Zahiri**
Fixes #2343

## Summary
In `alberta_framework/core/continual_backprop.py`:
`_select_replacement_index` ranked mature units by raw utility EMA rather than the bias-corrected utility defined in Eq. 8 of Dohare et al. (2024) ($\hat{u}_{l,i,t} = u_{l,i,t} / (1 - \eta^{a_{l,i,t}})$). Because replaced units reset their age and utility to 0, freshly matured units with short lifespans read at systematically lower raw EMA fractions ($\approx 1 - \eta^m \approx 0.634$ at $m=100, \eta=0.99$) compared to long-lived units ($\approx 1.0$), causing CBP to repeatedly target recently reset units.

## Proposed Changes
1. Updated `_select_replacement_index` to compute the per-unit bias correction $1 - \eta^{\text{age}}$ and rank eligible candidates by bias-corrected utility $\hat{u}$.
2. Forwarded `config.decay_rate` from `replace_units_with_flags`.
3. Added unit tests in `tests/test_continual_backprop.py` verifying that younger units with higher true contributions are preserved over older units with lower steady-state contributions.

## Verification
- Ran pytest: `/opt/homebrew/bin/uv run pytest -o pythonpath=. tests/test_continual_backprop.py` (96/96 passed).
- Ran ruff: `/opt/homebrew/bin/uv run ruff check alberta_framework/core/continual_backprop.py tests/test_continual_backprop.py` (clean).
- Ran mypy: `/opt/homebrew/bin/uv run mypy --follow-imports=silent alberta_framework/core/continual_backprop.py` (clean).

### CODE REVIEW

**Review by lumix5 — COMMENTED** [review]
**Disposition: close as superseded by #2760** — same CBP bias-corrected mature-unit ranking; #2760 is the surviving reviewed copy (current-head approval). Also duplicates #2533, itself recommended closed.

**Review by lumix5 — CHANGES_REQUESTED** [review]
Independent verification at the current head `c6311e70e7f2ee99a6151488f96ea13ffe8ea30c` (no prior review existed at any head).

**Disposition: fix** — as written this is an unmeasured algorithm change, not a defect repair.

- The module's own documented convention (`continual_backprop.py`, "Per-unit utility convention" and "Replacement" sections) states that selection picks "the mature unit with the lowest utility in the layer" from a plain EMA `u <- decay*u + (1-decay)*|a*g|`, and that the gradient form "is what the original Dohare implementation uses in practice". This PR silently changes that documented selection behavior.
- The citation does not check out. The official implementation ([shibhansh/loss-of-plasticity](https://github.com/shibhansh/loss-of-plasticity), `lop/algos/cbp_linear.py`) tracks utility as a plain EMA (`features = features * decay_rate + (1 - decay_rate) * input`) and selects via `torch.topk(-self.util[eligible_feature_indices], k)` — smallest **raw** utility among mature units, with no `1 - beta**t` correction anywhere in tracking or selection. The maturity threshold is the paper's mechanism for protecting young units ([Dohare et al., Nature 2024](https://www.nature.com/articles/s41586-024-07711-7); e.g. maturity 1,000 in Fig. 1b). If "Eq. 8" really defines selection-time bias correction, quote it verbatim in the PR body and update the module docstring — I could not find it in the paper or the reference code.
- `alberta_framework.core.continual_backprop` is on the import surface consumed by the robot track, so this behavior change propagates beyond this repository's lanes unmeasured.
- The new test proves the arithmetic of the correction, not that corrected selection improves any measured result here.

Path to acceptance: pre-register a paired comparison on a CBP-consuming lane (plain-EMA selection baseline vs bias-corrected selection, shared seeds, one variable, win threshold stated up front) and post the result — or have a maintainer explicitly bless the fidelity change, updating the "Replacement" docstring section and the citation. If the comparison loses, close and record the direction in `NEGATIVE_RESULTS_LEDGER.md`.

**Review by lumix5 — COMMENTED** [review]
**Reconciliation note from the same contributor** — two concurrent review sessions posted under this account ~18 minutes apart (the earlier close-as-superseded-by-#2760, the later CHANGES_REQUESTED). The later review's technical finding governs: the bias-correction contradicts the module docstring and the official `lop/algos/cbp_linear.py` (plain EMA, `topk(-util[eligible])`, no 1−β^t warm-up) — and **that finding transfers verbatim to #2760**, which proposes the same Dohare-style correction and currently carries an approval that predates this analysis. Neither #2569 nor #2760 should merge until the docstring/official-algorithm contradiction is resolved by measurement or the correction is withdrawn; the superseded/duplicate framing is moot while the underlying change is contested.

**Comment by lalalune** [comment]
Closing as superseded by merged #2345. The merged version applies per-unit EMA bias correction through the real replacement path and checks corrected utility eligibility. The replacement regression fails on main and the CBP/options panel passes all 206 tests.
